server/neptune/workflows/internal/pr/runner.go

```diff
@@ -90,30 +116,54 @@ func (r *Runner) Run(ctx workflow.Context) error {
 		c.Receive(ctx, &request)
 	})
 
+	onShutdownPollTick := func(f workflow.Future) {
+		action = onShutdownPoll
+	}
+	shutdownPollCancel, _ := s.AddTimeout(ctx, r.ShutdownPollTick, onShutdownPollTick)
+
+	_, revisionCancel := workflow.WithCancel(ctx)
 	for {
 		s.Select(ctx)
 		switch action {
 		case onNewRevision:
-			r.onNewRevision(ctx, prRevision)
+			revisionCancel = r.onNewRevision(ctx, revisionCancel, prRevision)
+			inactivityTimeoutCancel()
+			inactivityTimeoutCancel, _ = s.AddTimeout(ctx, r.InactivityTimeout, onInactivityTimeout)
+			continue
+		case onCancel:
+			continue
+		case onTimeout: // TODO: send message to PR stating atlantis deleted state due to inactivity and to rerun to trigger atlantis workflow
+			workflow.GetLogger(ctx).Info("workflow timed out, shutting down")
+			revisionCancel()
+			return nil
 		case onShutdown:
 			workflow.GetLogger(ctx).Info("received shutdown signal")
-			r.cancel()
-			return nil
+		case onShutdownPoll:
+			workflow.GetLogger(ctx).Info("shutdown check poll tick")
+		}
+		// shutdown check
+		shutdownPollCancel()
+		if shutdown := r.ShutdownChecker.ShouldShutdown(ctx, prRevision); !shutdown {
+			shutdownPollCancel, _ = s.AddTimeout(ctx, r.ShutdownPollTick, onShutdownPollTick)
+			continue
 		}
+		workflow.GetLogger(ctx).Info("pr is closed, shutting down")
+		r.state = complete
```

Comment:
Not sure if you should do this here. the go routine might set this to "waiting" after this happens.  Is there anything even checking completion state? Can we just not do this?

Reply:
Yeah it's not useful, nothing is checking for it so I'll remove it.